orders: add search overlay for manager orders

pajawojciech opened the pull request:
# Add search to manager orders

Adds search box to find orders by job name + arrow indicators to show which one you're looking at.

## What it does

- Search box appears on orders screen (Alt+S to focus)
- Type keywords to filter (e.g. "iron picks" finds "forge iron picks")
- Multi-word search works in any order ("picks iron" and "pi ir" also matches)
- Alt+P/N to jump between matches. Also Enter and Shift+Enter works when focused
- Shows "3 of 12" style counter
- Arrow point at current match (highlighted) and at all matches

## Implementation

- Uses df::interface_button_building_new_jobst to get order name
- New overlay: `OrdersSearchOverlay` enabled by default 
- Moved `importexport` overlay position to make room for search (and make new version of config)

## What didn't work

- Tried to clear search input when exiting orders window - couldn't find reliable way to detect window exit

## Related issues

- https://github.com/DFHack/dfhack/issues/1910
- https://github.com/DFHack/dfhack/issues/3346





https://github.com/user-attachments/assets/d4e5f219-2ecf-403b-8fdc-47896f5934c1

ab9rf commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +714,401 @@ function QuantityRightClickOverlay:onInput(keys)
+        order.mat_index,
+        order.reaction_name or '',
+        mat_cat_str,
+        encrust_str)
```
this results in lexicographic ordering by the text representation of job type's number, which means 0, 1, 10, 11, 12, ..., 19, 2, 20, 21...? while this is a stable sort order, it is not an order the player will likely anticipate

ChrisJohnsen commented on plugins/lua/orders.lua:
```diff
@@ -74,7 +79,7 @@ local mi = df.global.game.main_interface
 OrdersOverlay.ATTRS{
     desc='Adds import, export, and other functions to the manager orders screen.',
-    default_pos={x=53,y=-6},
+    default_pos={x=41,y=-6},
```
Without changing the overlay's `version`, changes to `default_pos` will only take effect for players that manually reset the overlay's position (e.g., `overlay position orders.importexport default` or use the "reset to default position" action (Ctrl+d) on the overlay in `gui/overlay`).

Note that a bumping the `version` will also cause the overlay to be automatically re-enabled for any player that had previous disabled it, so it isn't always appropriate.

Players that don't manually reset the position of `orders.importexport` will keep using their existing (possibly customized) position. If they are using the "old" default position, the `orders.importexport` window will overlap with the new `orders.search` window (which one is on top will vary by launch due to the vagaries of the Lua `pairs` iteration used to render the overlays).

pajawojciech replied:
Thanks, I bumped the version on importexport to force the new position.

ChrisJohnsen commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +714,401 @@ function QuantityRightClickOverlay:onInput(keys)
+    default_pos={x=85, y=-6},
+    default_enabled=false,
+    viewscreens='dwarfmode/Info/WORK_ORDERS/Default',
+    frame={w=34, h=4},
```
This combination of position and size does not properly fit in a minimum-size DF interface (114x46 UI cells, 912x552px with the default 8x12 UI font). The x-offset of 84 plus width of 34 puts the right edge at 118 columns.

This isn't a problem in "wide" DF windows, but if the the interface ends up narrower than 118 (due to window size and/or DF interface scaling percentage), then the `overlay.search` window will be pushed to the left. In a minimum-width interface this will cause the `orders.search` and `orders.importexport` windows (at their default positions established in this commit) to overlap by three columns (which one is on top varies across startups due to the Lua `pairs` iteration done for overlay rendering).

The leftward push experienced while the DF interface is narrower than 118 columns will remain until the overlay is reloaded (e.g., the player reloads the fort (via overlay's `onStateChange` handler), or `:lua require'plugins.overlay'.rescan()`). Only the widget's frame table is affected, not its persisted overlay configuration.

---

I don't know if it is standard procedure for DFHack (I'm just a contributor here), but I always try out overlay stuff in a variety of window sizes (especially a minimum-size window, which seems to heavily influences DF's own layout, and thus DFHack's DF-relative overlays) to make sure things will fit.

For this overlay, it looks like it could be narrowed by four columns without affecting its utility too much. Actually, I would recommend narrowing by six columns to keep the right edge away from the right edge of the (minimum-width) interface area (which will cause issues if `gui/overlay` is used to move the overlay window in a minimum-width interface: it will automatically swap to a right-anchored position which (in wider interfaces) will cause the overlay to "float" to the right instead of "sticking with" the (default position) of the `orders.importexport` overlay window).

pajawojciech replied:
I narrowed the overlay by 8 columns (now w=26) so it fits in minimum-size windows and still has room for 16 chars in the search field.

ChrisJohnsen commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +714,401 @@ function QuantityRightClickOverlay:onInput(keys)
+    return math.floor(available_height / self.ORDER_HEIGHT)
+end
+
+function OrderHighlightOverlay:calculateSelectedOrderY()
```
There is a bug here when deleting an order above the highlighted order. Here is a way to reproduce the problem:

1. Have more than one screen-full (page) of orders (i.e., the scrollbar is active).
1. Have a searchable order at or near the bottom of the the last page of orders.
1. Search for the order.
1. Move highlight to the order that is on the last page.

    The list should now be scrolled to the bottom of the list and there should be other orders above the highlighted order.

1. Delete an order above the highlighted order.

    The highlight is now drawn on the wrong row (or not drawn if it was "pushed off the bottom").

----

You might fix this by canceling the highlight (and invalidating the search -- or re-doing the search to update the match count) if the number of orders changes.

pajawojciech replied:
I added a check to hide the highlight when the order count changes

pajawojciech replied:
I want it to stay like this, header text will update on next search overlay user action

ChrisJohnsen commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +714,401 @@ function QuantityRightClickOverlay:onInput(keys)
+    return map
+end
+
+local reaction_map_cache = nil
```
I kept noticing a stutter when doing my first search of a bootup (I would type "plant" and end up with "pnt"). I think it is this cache being populated with ~3500 entries. Not sure if anything can be done to smooth that out. Maybe I am just running on a potato?

lethosor replied:
I would expect building the 3500 hash entries is probably the more expensive part, rather than inserting them into the hash. But it would be interesting to profile. I know you've been looking at lua table performance issues recently, but building a single hash in an application like this does not concern me much, at least from a runtime perspective, based partly on some simple "populate hash with 10000 elements"-type tests.

pajawojciech replied:
```
stockflow.collect_reactions() - 39 ms (runs once)
Building reaction map (3656 entries) - 141 ms (runs once)
get_order_search_key - under 1 ms (runs once per order, per search)
update_filter - 0-16 ms for 170 orders (runs once per search edit)
```

Based on these timings, the performance is good. We've got a ~180ms startup cost that only happens once. The actual searching is fast - under 16ms even with 170 orders. I can also cache the order keys, but I don't think it's worth it - search is already fast enough

pajawojciech commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +714,401 @@ function QuantityRightClickOverlay:onInput(keys)
+    desc='Shows arrows next to the work order found by orders.search',
+    default_enabled=false,
+    viewscreens='dwarfmode/Info/WORK_ORDERS/Default',
+    frame={w=80, h=3},
```
I changed it to full_interface=true and removed the frame since it's not needed.

pajawojciech commented:
>     * Would utils.search_text be useful here? The main SortOverlay uses it for searching.
I switched to using utils.search_text instead of the custom search logic
 
>     * Could the first match automatically be highlighted on Enter?
Added Enter/Shift+Enter to cycle through matches using the default submit/submit2 methods.

>     * The highlight should probably be cleared when the search text changes (especially when the highlighted order does not match the new search input).
Fixed, now the highlight gets cleared whenever the search text changes.

>     * It might just be my color vision being flaky, but I completely did not notice the highlight arrow at first. Now that I know what to look for, it isn't hard to spot. But my first cycling through the matches was confusing because it wasn't obvious which order was being indicated.
I reshaped the arrow, moved it to the right side of icon and used more contrasting colors (black on white) to make it more visible.
<img width="288" height="255" alt="obraz" src="https://github.com/user-attachments/assets/defe2178-b9fd-4289-af04-1ab85848ca43" />

---

@ChrisJohnsen Thanks for all the detailed feedback! I made each change in a separate commit to make the review easier. Let me know if there's anything else that needs adjusting.

ChrisJohnsen commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +716,397 @@ function QuantityRightClickOverlay:onInput(keys)
+    self.cached_list_start_y = nil
+    self.cached_viewport_size = nil
+    self.cached_screen_width = nil
+    self.cached_screen_height = nil
```
None of these appear to the used?

ChrisJohnsen commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +716,397 @@ function QuantityRightClickOverlay:onInput(keys)
+function OrderHighlightOverlay:render(dc)
+    if mi.job_details.open then return end
+
+    if search_cursor_visible then
```
The indentation isn't getting too bad here, but this might be better as a early-returning guard instead of having the rest of the body of the function tucked into the if block. It makes it clear that nothing else is going to happen for the guarded condition.

E.g.
```lua
if not search_cursor_visible then return end
```

ChrisJohnsen reviewed (commented):
Good spotting of the need to hide when `job_details.open`. The other functionality changes all seem good (with a new note about order deletion handling).

I've put some more thoughts on the specifics of the code in this review.

ChrisJohnsen commented:
So, I hadn't looked into the details of `collect_reactions` from `internal/quickfort/stockflow` until now... I had assumed (yeah, my mistake) that its `order` fields were Lua tables that mimicked `df::manager_order` objects enough that (e.g.) `make_order_key` could use either a real order or a "mocked up" one.

But those `order` fields are actual `df::manager_order` objects (~3500 of them) that are allocated and filled out in (functions called by) `collect_reactions` . So if callers don't free them, they end up leaked. Since these objects aren't needed by this code after keys are derived from them, they should probably just be freed right away.

The other caller is `internal/quickfort/orders`; it accepts the leak and only calls `collect_reactions` once (storing it in a module global; disclaiming any need for world-varying or modded reactions).

ChrisJohnsen commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +711,415 @@ function QuantityRightClickOverlay:onInput(keys)
+        encrust_str)
+end
+
+local reaction_map_cache = nil
```
When I tried loading different worlds I found that the cache wasn't being flushed like I expected. I mistakenly thought that a local cache variable would suffice. (I blame working too much on just overlays in script modules where a rescan will reload the module — when the script has been modified.)

The reaction cache needs to be flushed when a (new) world is loaded so that the world- or civilization-specific (e.g., `clothing_reactions`) orders can be cached.

For example, I have two forts: "A" knows how to make dresses but not shirts, "B" knows how to make shirts but not dresses. If I load "A" first, unload it, then load "B", searching "shirt" will not find a "make yarn shirt" order (though it does find still find a "forge iron mail shirt" order). This is because the "shirt" armor subtype was not cached since it was not available in fort "A".

Additionally, the cache should probably be a global variable. Leaving it as a local would force a rebuild after a module reload (even if the world hasn't changed).

---

You can watch for world loads with `dfhack.onStateChange`. The typical usage works like this:

At the top (after the `require` imports):

    local GLOBAL_KEY = 'orders'

The local can be replaced with a global initialization that copies any existing value instead of starting off as nil:

    reaction_map_cache = reaction_map_cache or nil

At the end (between `OVERLAY_WIDGETS` and `return _ENV`):

    dfhack.onStateChange[GLOBAL_KEY] = function(sc)
        if sc ~= SC_WORLD_LOADED then return end

        reaction_map_cache = nil
        get_cached_reaction_map()
    end

pajawojciech replied:
I changed my approach to name generation, so the comment is outdated. There is no cache now

ChrisJohnsen reviewed (commented):
I found couple of functionality things that I didn't notice before:
- cache invalidation between world loads,
- searchability after setting material for an order.

While looking at the details of customized orders I wondered if the overall size of the cache can be reduced.

Some functions can be made local (they don't access `self`, or won't once the methods the call are themselves made local):
- `perform_search`
- `getListStartY`
- `getViewportSize`
- `calculateSelectedOrderY`

When adding/removing an order there is still a small UI bug: the "Search: X of Y" title remains unchanged even if a matching order is added or removed. You could refresh the search when the number of orders changes, or just revert to the plain "Search" title so that the (possibly) stale count isn't displayed.

pajawojciech commented on plugins/lua/orders.lua:
```diff
@@ -709,11 +711,415 @@ function QuantityRightClickOverlay:onInput(keys)
+local search_last_scroll_position = -1
+local order_count_at_highlight = 0
+
+local function make_order_key(order)
```
Thank you for this proposition. 
I changed my approach to name generation, so now there is no keys

ChrisJohnsen reviewed (commented):
Here is a review to tick the requested-review box. My previous comment also mentions a couple of remaining UI bugs.

ChrisJohnsen commented:
Found a leftover while reviewing my local notes:

If `getManagerOrderName` ends up in the C++-exported Lua API (whether it ends up in Jobs or elsewhere), it should probably get an entry in `docs/dev/Lua API.rst` (similar to the one for `dfhack.job.getName`).

pajawojciech commented on library/modules/Job.cpp:
```diff
@@ -686,3 +687,29 @@ std::string Job::getName(df::job *job)
+    button->specflag = order->specflag;
+    button->job_item_flag = order->material_category;
+    button->specdata = order->specdata;
+    button->art_specifier = order->art_spec.type;
```
It is not needed. Removed

pajawojciech commented on library/modules/Job.cpp:
```diff
@@ -686,3 +687,29 @@ std::string Job::getName(df::job *job)
+    std::string desc;
+    auto button = df::allocate<df::interface_button_building_new_jobst>();
+    button->mstring = order->reaction_name;
+    button->specdata.hist_figure_id = order->specdata.hist_figure_id;
```
It is a struct and there is no need to copy hist_figure_id. Removed

pajawojciech commented:
> Here is a review to tick the requested-review box.

Sorry, but I don't understand what this is about.


> My previous comment also mentions a couple of remaining UI bugs.

UI bugs fixed and some more changes:
- render method now contains periodically check for orders changes
- search automatically refreshes when orders are added, removed, or modified
- all matched orders are highlighted, not just the selected one
- matching orders are highlighted in real-time as you type
- highlights remain visible while scrolling for selected and matching orders

I love how it feels to use now that the order highlights stay visible. It's much more user-friendly. Thanks @ChrisJohnsen for the review!

<img width="849" height="514" alt="obraz" src="https://github.com/user-attachments/assets/fc7948cb-6969-442b-b0c7-22f174339e74" />

ChrisJohnsen commented:
> > Here is a review to tick the requested-review box.
> 
> Sorry, but I don't understand what this is about.

Yeah, that was just GitHub noise. My [first comment][a] ("Interesting. …") was just a "comment" (not a "review"), so it did not clear the "review requested" state. I posted the [second one][b] ("Here is a review…") as a actual "review" to have GitHub mark the PR as "reviewed".

[a]: https://github.com/DFHack/dfhack/pull/5677#issuecomment-3703699680
[b]: https://github.com/DFHack/dfhack/pull/5677#pullrequestreview-3621870537

ChrisJohnsen commented:
Since all matches are now being highlighted `search_cursor_visible` is now synonymous with `#search_matched_indices > 0`, so a separate variable isn't really necessary.

---

You might want to use time-based updating instead of frame-based so that the update time doesn't vary (as much) with configured (or effective) frame rate.

The overlay system provides periodic calls to the `overlay_onupdate` method when it is defined. The period can be controlled by the `overlay_onupdate_max_freq_seconds` attribute/field. The default period is 5 seconds, but since the game is paused while these overlays are active it should be okay to request more frequent updates. A 1 second period would approximate your 50 frame interval (for the default 50 FPS setting), but float values between 0 and 1 also work here and could make it more responsive.

Hmm, this is another case where a "post input" callback  would be handy. I've come across other uses for an overlay callback that occurs _after_ DF has handled an input. Pretty much all player-originated order changes will need to done through some kind of input, so this overlay could (if a "post input" callback was available) just check whether the orders have changed after each input instead of periodically... I've added this use case to my notes.

---

Since the inter-overlay coordination has been awkward a couple of times (and just changed to be bi-directional with `search_overlay_instance`), it might be worth considering consolidating the overlays. I have shied away from suggesting it since it involves a bit of non-standard `render` usage, but it might end up being less awkward than the variable-mediated inter-overlay coordination.

The variable-mediated coordination relies to the pseudo-singleton nature of overlays (as managed by the overlay system) which isn't documented at all as far as I have seen.

Drawing the highlights outside the main widget's frame is kind of non-standard behavior and relies on the painter being clipped to the full interface area.

The `view:render` documentation does state that the parent's painter is supplied to `render`, but it is an assumption that an overlay widget's parent's painter is clipped to the full interface area (though not much of an assumption, in my opinion, since overlay widgets can normally be placed anywhere in the interface area). `render` is already present to cut off drawing when `mi.job_details.open`, so it would be easy to call a "draw highlights" function from there.

Neither is a perfect, but combining the overlays would get rid of all the non-constant variables.

---

A minor thing about "checksum": To me, the "checksum" strongly implies a small (or constant-sized) value. The concatenation of order names is serving a similar purpose (change detection), but it is basically just a copy of the data it is checking. Something like `last_order_names` might be better?

I experimented with a "checksum"-free version that just periodically, unconditionally called a modified `update_filter` that checked whether the new `search_matched_indices` was different from the old value and only reset `search_current_match_idx` when there was a change. It worked okay, but it did pull the frame rate down a bit (this was with a 0.1 second period, so fairly aggressive). The idea was that `perform_search` needed to get all the order names anyway, maybe that work could also serve to detect changes. But it looks like all the `utils.search_text` usage is more expensive than just collecting the order names an extra time.

ab9rf commented on docs/dev/Lua API.rst:
```diff
@@ -1430,6 +1430,10 @@ Job module
 
   Returns the job's description, as seen in the Units and Jobs screens.
 
+* ``dfhack.job.getManagerOrderName(manager_order)``
```
This API change requires a changelog entry (in the API section).

pajawojciech replied:
Added

ab9rf commented on docs/changelog.txt:
```diff
@@ -115,6 +116,7 @@ Template for new versions:
 ## Documentation
 
 ## API
+- `dfhack.job.getManagerOrderName()`: New function to get the display name of a manager order
```
need doubled backticks here

pajawojciech replied:
Done
